File: athleteiq/web/static/ball.js
/**
 * Ball tracking, on the phone, alongside pose.
 *
 * The rest of this product counts what the *body* did. A wall-ball rep is a
 * throwing motion, a squat is a knee angle -- which is why the catalog has no
 * dribbling or juggling drills, and why the README says a convincing
 * shadow-throw with no ball counts. This module is what closes that.
 *
 * Three things shape it.
 *
 * **It has to run next to pose estimation on a mid-range phone.** A detector
 * heavy enough to find a tennis ball reliably is far too slow to run every
 * frame while MediaPipe is already using the GPU. So detection runs at a low
 * rate and a cheap predictive tracker fills the gaps -- the standard
 * detect-then-track split. Frames the tracker invented are labelled as such and
 * never counted as evidence.
 *
 * **A contact is physics, not a heuristic.** A ball in free flight accelerates
 * downward at a constant rate; anything else is something hitting it. So the
 * detector estimates gravity *from the track itself* -- which makes it
 * independent of how far away the camera is, what the frame height represents,
 * and whether the phone is in portrait -- and flags frames where the velocity
 * change is far larger than gravity explains. That is an impulse, and an
 * impulse is a foot, a floor, or a wall.
 *
 * **When it cannot see the ball it must say so.** Every track reports the
 * share of frames backed by a real observation. A drill that needs a ball
 * refuses to count below the floor rather than degrading quietly into a pose
 * drill that says "42 juggles" when the athlete was standing still. Fabricated
 * counts are worse than no feature.
 */

import { LANDMARKS } from './counter.js';

/**
 * Distance between two normalised points, corrected for frame shape.
 *
 * Pose landmarks and ball positions both arrive normalised 0-1 against their
 * own axis, which makes that space **anisotropic**: in a 16:9 landscape frame
 * one x-unit is 1.78 times wider on the ground than one y-unit, and in
 * portrait it is the other way round. Every radius and gate in this module is
 * a real-world distance, so measuring them in raw normalised units meant a
 * phone turned sideways silently changed what "next to the ball" meant.
 *
 * Everything here is therefore measured in **frame heights**: x is scaled by
 * the aspect ratio, y is left alone, and a distance means the same thing
 * whichever way up the phone is.
 */
export function metricDistance(ax, ay, bx, by, aspect = 1) {
  return Math.hypot((ax - bx) * aspect, ay - by);
}

/** Detections older than this are not worth extrapolating from. */
export const MAX_COAST_MS = 220;

/** Search radius for associating a detection with the existing track. */
export const GATE_RADIUS = 0.18;

/**
 * Consecutive out-of-gate detections before the track gives up and re-acquires.
 *
 * Without this the tracker locks onto whatever it saw first and can never
 * recover -- a ball that leaves the frame and comes back somewhere else, or a
 * first frame that caught a ball lying on the grass behind the athlete, would
 * make the whole session count nothing while the real ball was rejected as a
 * decoy for being too far from a track that was wrong to begin with.
 */
export const REACQUIRE_AFTER = 5;

/** Prior for downward acceleration, in frame-heights per second squared. */
export const GRAVITY_PRIOR = 1.9;

/** An impulse this much larger than gravity predicts is a contact. */
export const IMPULSE_RATIO = 3.0;

/** Minimum outgoing speed for a contact to be real rather than a wobble. */
export const MIN_CONTACT_SPEED = 0.25;

const clamp = (v, lo, hi) => Math.max(lo, Math.min(hi, v));

/**
 * A smoothed ball track built from sparse detections.
 *
 * Deliberately not a full Kalman filter: the state is two-dimensional and the
 * measurement noise is dominated by the detector's own box jitter, so a
 * constant-velocity model with an exponential blend behaves the same and is
 * small enough to read.
 */
export class BallTracker {
  constructor(options = {}) {
    this.gate = options.gate ?? GATE_RADIUS;
    this.reacquireAfter = options.reacquireAfter ?? REACQUIRE_AFTER;
    //: Frame width over height. 1 until the camera reports otherwise.
    this.aspect = options.aspect ?? 1;
    this.maxCoastMs = options.maxCoastMs ?? MAX_COAST_MS;
    this.positionBlend = options.positionBlend ?? 0.55;
    this.velocityBlend = options.velocityBlend ?? 0.45;
    this.reset();
  }

  reset() {
    this.x = null; this.y = null;
    this.vx = 0; this.vy = 0;
    this.r = 0.02;
    this.lastSeenAt = null;
    this.lastAt = null;
    /** The previous *observation*, which is what velocity is measured across. */
    this.lastObs = null;
    this.observed = 0;
    this.frames = 0;
    this.rejected = 0;
    /** Running estimate of gravity in this framing, learned from free flight. */
    this.gravity = GRAVITY_PRIOR;
    this.gravitySamples = 0;
    this.lastObsVy = null;
    /** Raw velocity between the last two observations, before smoothing. */
    this.obsVx = 0;
    this.obsVy = 0;
  }

  /** Share of frames that had a real detection behind them. */
  get quality() {
    return this.frames ? this.observed / this.frames : 0;
  }

  get visible() {
    return this.x !== null;
  }

  /**
   * Fold one frame in.
   *
   * `detection` is `{x, y, r, score}` in normalised frame coordinates, or null
   * on frames where the detector did not run or found nothing.
   */
  push(detection, tMs) {
    const dt = this.lastAt === null ? 0 : (tMs - this.lastAt) / 1000;
    this.lastAt = tMs;
    this.frames += 1;

    if (detection) {
      const outOfGate = this.visible
        && metricDistance(detection.x, detection.y, this.x, this.y, this.aspect) > this.gate
        && (tMs - this.lastSeenAt) < this.maxCoastMs;

      // A detection far from where the ball must be, while the track is still
      // fresh, is another object -- a face, a shoe, a ball on the ground
      // behind the athlete. Ignoring it keeps the track on the real ball.
      // But a run of them means the track itself is the thing that is wrong.
      if (outOfGate && this.rejected < this.reacquireAfter) {
        this.rejected += 1;
      } else {
        if (outOfGate) this._reacquire();
        this.rejected = 0;
        this._observe(detection, tMs);
        return this.state('detect');
      }
    }

    if (!this.visible) return this.state('none');

    if (tMs - this.lastSeenAt > this.maxCoastMs) {
      // Coasted too long. Dropping the track is honest; extrapolating a ball
      // through a second of nothing invents a trajectory.
      this.x = null; this.y = null;
      return this.state('none');
    }

    // Coast on the constant-velocity model, with gravity applied so a ball
    // mid-flight lands where physics says rather than where it was going.
    this.x += this.vx * dt;
    this.y += this.vy * dt + 0.5 * this.gravity * dt * dt;
    this.vy += this.gravity * dt;
    return this.state('predict');
  }

  /** Forget the current track so the next detection starts a fresh one. */
  _reacquire() {
    this.x = null; this.y = null;
    this.vx = 0; this.vy = 0;
    this.lastObs = null;
    this.lastObsVy = null;
  }

  _observe(detection, tMs) {
    // Velocity is measured between observations, never between frames. The
    // difference matters: with the detector at 10fps and the camera at 30,
    // `this.x` has already been coasted forward twice using the current
    // gravity estimate, so dividing by the frame interval measures the
    // tracker's own prediction back to itself. That feedback loop inflated
    // the learned gravity by 50% and manufactured phantom contacts.
    const obs = this.lastObs;
    const dtObs = obs ? (tMs - obs.t) / 1000 : 0;

    if (!obs || dtObs <= 0 || dtObs > this.maxCoastMs / 1000) {
      this.x = detection.x; this.y = detection.y;
      this.vx = 0; this.vy = 0;
      this.obsVx = 0; this.obsVy = 0;
    } else {
      const vx = (detection.x - obs.x) / dtObs;
      const vy = (detection.y - obs.y) / dtObs;
      // Kept unsmoothed. Smoothing exists so the drawn ball does not jitter;
      // running impulse detection on it damps the very spike being looked for
      // -- at a 10fps detection rate that hid eight kicks in nine.
      this.obsVx = vx;
      this.obsVy = vy;
      this._learnGravity(vy, dtObs);
      this.vx += this.velocityBlend * (vx - this.vx);
      this.vy += this.velocityBlend * (vy - this.vy);
      this.x += this.positionBlend * (detection.x - this.x);
      this.y += this.positionBlend * (detection.y - this.y);
    }
    this.lastObs = { x: detection.x, y: detection.y, t: tMs };
    this.r = detection.r ?? this.r;
    this.lastSeenAt = tMs;
    this.observed += 1;
  }

  /**
   * Learn how fast things fall in *this* framing.
   *
   * A phone held close sees a ball accelerate across far more of the frame per
   * second than one filming from the sideline. Estimating it from the athlete's
   * own footage means the contact detector never needs to know the camera
   * distance, the frame height in metres, or the orientation.
   */
  _learnGravity(vy, dt) {
    if (dt <= 0 || dt > 0.2) return;
    if (this.lastObsVy === undefined || this.lastObsVy === null) {
      this.lastObsVy = vy;
      return;
    }
    const accel = (vy - this.lastObsVy) / dt;
    this.lastObsVy = vy;
    // Only free-flight frames teach anything. A contact is a huge acceleration
    // and would drag the estimate up until nothing looked like a contact.
    if (accel <= 0 || accel > 12) return;
    this.gravitySamples += 1;
    const weight = 1 / Math.min(this.gravitySamples, 40);
    this.gravity += weight * (accel - this.gravity);
    this.gravity = clamp(this.gravity, 0.4, 8);
  }

  state(source) {
    return {
      x: this.x, y: this.y, r: this.r,
      vx: this.vx, vy: this.vy,
      obsVx: this.obsVx, obsVy: this.obsVy,
      speed: Math.hypot(this.vx, this.vy),
      obsSpeed: Math.hypot(this.obsVx, this.obsVy),
      source,
      quality: this.quality,
      gravity: this.gravity,
    };
  }
}

/**
 * Turns a ball track plus pose landmarks into contacts.
 *
 * A contact is a velocity change gravity cannot explain. Everything else here
 * is about deciding *what* hit it, which is what makes a juggle different from
 * a dribble.
 */
export class ContactDetector {
  constructor(options = {}) {
    this.minGapMs = options.minGapMs ?? 180;
    this.impulseRatio = options.impulseRatio ?? IMPULSE_RATIO;
    this.minSpeed = options.minSpeed ?? MIN_CONTACT_SPEED;
    this.partRadius = options.partRadius ?? 0.16;
    this.groundBand = options.groundBand ?? 0.12;
    this.aspect = options.aspect ?? 1;
    this.reset();
  }

  reset() {
    this.prev = null;
    this.prevAt = null;
    this.lastContactAt = -Infinity;
    this.contacts = [];
    this.seen = 0;
  }

  /**
   * Feed one tracked frame. Returns a contact when this frame was one.
   *
   * `parts` is a map of name -> {x, y} for the landmarks this drill cares
   * about, already scaled into the same normalised space as the ball.
   */
  push(ball, parts, tMs) {
    if (!ball || ball.x === null) { this.prev = null; return null; }

    // Contacts are only read off frames with a real detection behind them.
    // On a coasted frame the velocity is whatever the model extrapolated, and
    // on the frame an observation lands it takes a corrective step -- which
    // looks exactly like an impulse. Requiring evidence removes both, and is
    // the same rule the rest of this product follows: a count has to be backed
    // by something the camera actually saw.
    if (ball.source !== 'detect') return null;
    this.seen += 1;

    const prev = this.prev;
    const prevAt = this.prevAt;
    this.prev = { vx: ball.obsVx, vy: ball.obsVy, x: ball.x, y: ball.y };
    this.prevAt = tMs;
    if (!prev || prevAt === null) return null;

    const dt = (tMs - prevAt) / 1000;
    if (dt <= 0 || dt > 0.25) return null;
    // The first measured velocity is compared against a standing start of
    // zero, so a ball already in flight when recording began reads as a
    // contact. Two observations are needed before any of this means anything.
    if (this.seen < 3) return null;

    // What the velocity would be if nothing had touched it.
    const expectedVy = prev.vy + ball.gravity * dt;
    const residual = Math.hypot(ball.obsVx - prev.vx, ball.obsVy - expectedVy);
    const explained = Math.abs(ball.gravity * dt);

    if (residual < Math.max(this.minSpeed, explained * this.impulseRatio)) return null;
    if (tMs - this.lastContactAt < this.minGapMs) return null;
    // A ball leaving a contact is going somewhere. This drops the shivers a
    // jittery detection box produces when the ball is nearly still.
    if (ball.obsSpeed < this.minSpeed) return null;

    const contact = {
      t: tMs,
      x: ball.x, y: ball.y,
      impulse: residual,
      speedIn: Math.hypot(prev.vx, prev.vy),
      speedOut: ball.obsSpeed,
      ...this.classify(ball, parts),
    };
    this.lastContactAt = tMs;
    this.contacts.push(contact);
    return contact;
  }

  /** Who or what hit it: the nearest listed landmark, or the floor. */
  classify(ball, parts) {
    let best = null;
    let bestDistance = Infinity;
    for (const [name, point] of Object.entries(parts || {})) {
      if (!point) continue;
      const distance = metricDistance(ball.x, ball.y, point.x, point.y, this.aspect);
      if (distance < bestDistance) { bestDistance = distance; best = name; }
    }

    // Near the bottom of the frame with nothing close by is the floor. Checked
    // after the landmarks so a foot at ground level still reads as a foot,
    // which is the whole difference between a dribble and a juggle.
    if (best !== null && bestDistance <= this.partRadius) {
      return {
        kind: 'body',
        part: best,
        side: best.startsWith('left_') ? 'left'
          : best.startsWith('right_') ? 'right' : 'none',
        distance: bestDistance,
      };
    }
    if (ball.y > 1 - this.groundBand) {
      return { kind: 'ground', part: 'ground', side: 'none', distance: bestDistance };
    }
    return { kind: 'other', part: '', side: 'none', distance: bestDistance };
  }
}

/**
 * Turns ball contacts into reps, per a drill's BallSpec.
 *
 * Deliberately a separate counter from `RepCounter` rather than another branch
 * inside it. The pose counter's whole model is a smoothed signal crossing
 * hysteresis thresholds; a ball rep is a discrete event with no signal and no
 * range of motion. Forcing them together would mean a threshold field that
 * means nothing for half the catalog.
 */
export class BallRepCounter {
  constructor(spec) {
    this.spec = spec;
    this.ball = spec.ball || {};
    this.tracker = new BallTracker();
    this.detector = new ContactDetector({
      minGapMs: this.ball.min_gap_ms ?? 180,
      minSpeed: this.ball.min_speed ?? MIN_CONTACT_SPEED,
    });
    this.aspect = 1;
    this.reset();
  }

  /**
   * Tell the counter what shape the frame is.
   *
   * Set once the camera reports its dimensions, and again if the phone is
   * turned mid-session -- which athletes do, and which used to change the
   * meaning of every distance in here.
   */
  setAspect(aspect) {
    if (!aspect || !Number.isFinite(aspect) || aspect <= 0) return;
    this.aspect = aspect;
    this.tracker.aspect = aspect;
    this.detector.aspect = aspect;
  }

  reset() {
    this.tracker.reset();
    this.detector.reset();
    this.reps = [];
    this.lastAt = null;
    /** Frames where the ball was tracked at all. */
    this.trackedFrames = 0;
    /**
     * Frames where the ball was well clear of both hands.
     *
     * The check that separates wall ball from waving a ball around. An arm
     * whipping through a throwing motion with the ball still in it produces
     * accelerations that read as impulses and a wrist right beside them that
     * reads as a contact -- so contact counting alone scores a fake exactly
     * like the real thing. What it cannot fake is the ball leaving: real wall
     * ball sends it metres away and brings it back, and a ball that never
     * gets more than a hand's width from a wrist never went anywhere.
     */
    this.awayFrames = 0;
  }

  get count() { return this.reps.length; }

  get trackQuality() { return this.tracker.quality; }

  /** Whether this session is trustworthy enough to count at all. */
  get trusted() {
    return this.trackQuality >= (this.ball.min_track_quality ?? 0.35);
  }

  handCounts() {
    let left = 0, right = 0;
    for (const rep of this.reps) {
      if (rep.hand === 'left') left += 1;
      else if (rep.hand === 'right') right += 1;
    }
    return { left, right };
  }

  /**
   * One frame. `detection` may be null on frames the detector skipped;
   * `landmarks` is the pose array, or null if pose has not resolved yet.
   */
  push(detection, landmarks, tMs) {
    this.lastAt = tMs;
    const ball = this.tracker.push(detection, tMs);
    this._measureTravel(ball, landmarks);
    const contact = this.detector.push(ball, this.parts(landmarks), tMs);
    if (!contact) return null;

    // A juggle only counts if a body part took it; a dribble only counts if
    // the floor did. Without this a ball bouncing off a wall behind the
    // athlete would count as a touch.
    const wanted = this.ball.contact ?? 'body';
    if (contact.kind !== wanted) return null;

    const rep = {
      t_ms: Math.round(contact.t),
      hand: this.ball.attribute_side ? sideOf(contact, landmarks, this.aspect) : 'none',
      confidence: Math.min(1, this.tracker.quality + 0.2),
      part: contact.part,
      speed: Number(contact.speedOut.toFixed(3)),
      impulse: Number(contact.impulse.toFixed(3)),
    };
    this.reps.push(rep);
    return rep;
  }

  /** How far the ball gets from the hands, measured in the athlete's torsos. */
  _measureTravel(ball, landmarks) {
    if (!ball || ball.x === null || !landmarks) return;
    const torso = torsoLength(landmarks, this.aspect);
    if (!torso) return;
    this.trackedFrames += 1;

    let nearest = Infinity;
    for (const name of ['left_wrist', 'right_wrist']) {
      const point = landmarks[LANDMARK_INDEX[name]];
      if (!point || (point.visibility ?? 1) < 0.4) continue;
      nearest = Math.min(
        nearest, metricDistance(ball.x, ball.y, point.x, point.y, this.aspect),
      );
    }
    if (nearest === Infinity) return;
    if (nearest / torso > AWAY_FROM_HANDS) this.awayFrames += 1;
  }

  /** Share of tracked frames where the ball was genuinely away from the hands. */
  get travelShare() {
    return this.trackedFrames ? this.awayFrames / this.trackedFrames : 0;
  }

  /**
   * Corroboration for a drill whose reps the *body* counted.
   *
   * Returned as fields to merge into the pose counter's submission rather
   * than a submission of its own: in confirm mode the reps belong to the pose
   * counter and this only ever adds to them.
   */
  confirmation() {
    return {
      track_quality: Math.round(this.trackQuality * 1000) / 1000,
      ball_contacts: this.detector.contacts.length,
      ball_travel: Math.round(this.travelShare * 1000) / 1000,
    };
  }

  /**
   * The payload posted to /api/sessions/submit. Counts only, same as the pose
   * counter -- plus the track quality, which the server checks against the
   * drill's own floor rather than taking the client's word for.
   */
  toSubmission(sessionId, nonce, durationMs, extra = {}) {
    return {
      session_id: sessionId,
      nonce,
      duration_ms: Math.round(durationMs),
      reps: this.reps,
      hold_ms: 0,
      mean_confidence: Math.round(this.trackQuality * 1000) / 1000,
      track_quality: Math.round(this.trackQuality * 1000) / 1000,
      ...extra,
    };
  }

  /** The landmarks this drill treats as able to touch the ball. */
  parts(landmarks) {
    const out = {};
    if (!landmarks) return out;
    for (const name of this.ball.parts || []) {
      const index = LANDMARK_INDEX[name];
      const point = index === undefined ? null : landmarks[index];
      if (point && (point.visibility ?? 1) > 0.4) {
        out[name] = { x: point.x, y: point.y };
      }
    }
    return out;
  }
}

/** Distance from ball to hand, beyond which the ball has actually gone. */
export const AWAY_FROM_HANDS = 1.5;

/** Shoulder-to-hip, the scale everything about a body is measured in. */
function torsoLength(landmarks, aspect = 1) {
  const shoulder = landmarks[LANDMARK_INDEX.left_shoulder]
    || landmarks[LANDMARK_INDEX.right_shoulder];
  const hip = landmarks[LANDMARK_INDEX.left_hip] || landmarks[LANDMARK_INDEX.right_hip];
  if (!shoulder || !hip) return 0;
  const torso = metricDistance(shoulder.x, shoulder.y, hip.x, hip.y, aspect);
  return torso > 0.05 ? torso : 0;
}

/**
 * Which side took the contact.
 *
 * For a body contact the landmark says it directly. For a ground contact --
 * a dribble -- the floor has no side, so it is the nearer wrist that answers.
 */
function sideOf(contact, landmarks, aspect = 1) {
  if (contact.side && contact.side !== 'none') return contact.side;
  if (!landmarks) return 'none';
  const left = landmarks[LANDMARK_INDEX.left_wrist];
  const right = landmarks[LANDMARK_INDEX.right_wrist];
  if (!left || !right) return 'none';
  const dl = metricDistance(contact.x, contact.y, left.x, left.y, aspect);
  const dr = metricDistance(contact.x, contact.y, right.x, right.y, aspect);
  // Too close to call is 'none' rather than a coin flip: a fabricated
  // left/right split would feed straight into the off-hand balance score.
  if (Math.abs(dl - dr) < 0.05) return 'none';
  return dl < dr ? 'left' : 'right';
}

/**
 * Landmark name to array index.
 *
 * Imported from the counter rather than injected by the caller. The injected
 * version was one forgotten call away from an empty map, which would not
 * throw -- it would silently classify every contact as "nothing was near it"
 * and count zero juggles while reporting a healthy track.
 */
export const LANDMARK_INDEX = Object.fromEntries(
  LANDMARKS.map((name, index) => [name, index]),
);

File: offdays/ball.py
"""Server-side checks on ball-tracked sessions.

The browser does the tracking, so the browser is where the numbers come from,
so the browser cannot be trusted with them -- the same reasoning that produced
`integrity.py` for pose drills. A ball payload is easier to fake than a pose
one, because a rep is just a timestamp and a speed rather than a whole skeleton.

Three things are checked here that the client cannot be relied on to enforce.

**Track quality.** Every ball drill declares the share of frames that must
have had a real detection behind them. A client can report whatever it likes;
the server compares it against the spec and holds the session for review if it
is short. Silence is the failure mode being avoided: a session that says
"42 juggles" from a track that saw the ball twice.

**Physical plausibility.** Contacts have a rhythm. Juggling is roughly one
touch a second because the ball has to come back down; a payload claiming
forty touches in four seconds is describing something that did not happen.
Gaps below the drill's own refractory window cannot occur in a real client at
all, so their presence means the payload was not produced by one.

**Left/right honesty.** A perfectly even split across hundreds of contacts is
not what a child juggling in a garden produces. It is what a loop produces.
"""

from __future__ import annotations

import statistics
from dataclasses import dataclass, field
from typing import Any

from .drills.base import DrillSpec

#: Below this the session is held rather than counted. A separate, lower bar
#: than the drill's own floor: the client should have stopped already, so
#: reaching the server at all means something is off.
HARD_QUALITY_FLOOR = 0.15

#: A left/right split this close to perfect, over this many reps, is a loop.
SUSPICIOUS_BALANCE = 0.002
BALANCE_MIN_REPS = 40

#: Confirm mode. Below this share of reps corroborated by a ball contact --
#: on a session where the ball *was* clearly visible -- the reps did not
#: involve a ball.
CONFIRM_MIN_SHARE = 0.25

#: And not judged at all below this many reps: a handful of throws where the
#: detector happened to miss the ball is noise, not evidence.
CONFIRM_MIN_REPS = 15

#: Share of tracked frames the ball must spend away from the hands.
#:
#: The check that separates wall ball from waving a ball around. An arm
#: whipping through a throwing motion with the ball still in it produces
#: accelerations that read as impulses and a wrist beside them that reads as a
#: contact, so contact counting alone scores that fake exactly like the real
#: thing -- measured at twelve contacts for twelve fake throws. What it cannot
#: fake is the ball leaving: real wall ball sends it metres away and brings it
#: back, and a ball that never gets more than a hand's width from a wrist has
#: not been thrown at anything.
CONFIRM_MIN_TRAVEL = 0.15


@dataclass
class BallReview:
    ok: bool = True
    hold: bool = False
    reasons: list[str] = field(default_factory=list)
    quality: float = 0.0

    #: Things worth saying that are not accusations. Kept separate from
    #: `reasons` so a note can never hold a session by accident.
    notes: list[str] = field(default_factory=list)

    def flag(self, reason: str, hold: bool = True) -> None:
        self.reasons.append(reason)
        self.ok = False
        self.hold = self.hold or hold

    def note(self, message: str) -> None:
        self.notes.append(message)

    def to_dict(self) -> dict[str, Any]:
        return {
            "ok": self.ok,
            "hold": self.hold,
            "reasons": list(self.reasons),
            "notes": list(self.notes),
            "quality": round(self.quality, 3),
        }


def review(
    spec: DrillSpec,
    reps: list[dict[str, Any]],
    track_quality: float | None,
    duration_ms: int,
    ball_contacts: int | None = None,
    ball_travel: float | None = None,
) -> BallReview:
    """Decide whether a ball-tracked submission can be counted."""
    result = BallReview(quality=float(track_quality or 0.0))
    if spec.ball is None:
        return result
    if spec.ball.confirms:
        return _review_confirmation(
            spec, reps, track_quality, ball_contacts, ball_travel, result,
        )

    if track_quality is None:
        result.flag(
            "This drill needs the ball tracked, and the recording did not "
            "report whether it saw one."
        )
        return result

    floor = spec.ball.min_track_quality
    if result.quality < HARD_QUALITY_FLOOR:
        result.flag(
            f"The ball was only visible in {result.quality:.0%} of the frames. "
            "Try again with the ball and your whole body in shot."
        )
    elif result.quality < floor:
        result.flag(
            f"The ball was visible in {result.quality:.0%} of frames, under the "
            f"{floor:.0%} this drill needs to count."
        )

    if not reps:
        return result

    times = sorted(int(r.get("t_ms", 0)) for r in reps)
    gaps = [b - a for a, b in zip(times, times[1:])]

    # A real client enforces the refractory window itself, so a gap below it
    # did not come from one.
    too_fast = [g for g in gaps if g < spec.ball.min_gap_ms * 0.75]
    if too_fast:
        result.flag(
            f"{len(too_fast)} contacts came faster than this drill allows.",
        )

    if duration_ms > 0:
        rate = len(reps) / (duration_ms / 1000)
        if rate > spec.validation.max_reps_per_second:
            result.flag(
                f"{rate:.1f} contacts a second is faster than the ball can "
                "come back."
            )

    # Perfectly even timing is a generator, not a child in a garden. Same
    # check the pose integrity layer makes, for the same reason.
    if len(gaps) >= 12:
        mean = statistics.fmean(gaps)
        if mean > 0:
            spread = statistics.pstdev(gaps) / mean
            if spread < 0.03:
                result.flag("The contacts are too evenly spaced to be real.")

    _review_alternation(spec, reps, result)

    if spec.ball.attribute_side:
        left = sum(1 for r in reps if r.get("hand") == "left")
        right = sum(1 for r in reps if r.get("hand") == "right")
        total = left + right
        if total >= BALANCE_MIN_REPS:
            share = left / total
            if abs(share - 0.5) < SUSPICIOUS_BALANCE:
                result.flag(
                    "The left/right split is too exact to have happened.",
                )
    return result


#: Below this many attributed contacts the hand pattern is noise. A crossover
#: session of six bounces says nothing about whether the athlete crossed over.
ALTERNATION_MIN_CONTACTS = 12

#: A crossover drill is not required to be perfect -- a fumbled rep or a hand
#: the camera misread should not fail the session. It is required to look like
#: somebody changing hands rather than somebody dribbling on one.
ALTERNATION_FLOOR = 0.60

#: ...and the mirror of it: a one-handed drill may lose the odd contact to
#: misattribution without becoming a two-handed one.
SAME_HAND_FLOOR = 0.80


def _review_alternation(
    spec: DrillSpec, reps: list[dict[str, Any]], result: BallReview,
) -> None:
    """Check the hands did what the drill asked of them.

    The point of this check is what it lets the catalogue do honestly. A
    crossover pays more than a plain dribble, and that is only defensible
    because this function can tell the difference -- unlike the wall-ball
    patterns, where the fancier name paid more for a movement the camera could
    not distinguish at all.

    Worded as a note rather than a refusal when the pattern is simply absent:
    an athlete who meant to cross over and mostly dribbled on their strong hand
    has done real work and should be told what happened, not have the session
    thrown away.
    """
    rule = spec.ball.alternation
    if rule == "any":
        return
    hands = [r.get("hand") for r in reps if r.get("hand") in ("left", "right")]
    if len(hands) < ALTERNATION_MIN_CONTACTS:
        return

    swaps = sum(1 for a, b in zip(hands, hands[1:]) if a != b)
    share = swaps / (len(hands) - 1)

    if rule == "alternating" and share < ALTERNATION_FLOOR:
        result.note(
            f"The ball changed hands on {share:.0%} of contacts. This drill is "
            "the change itself -- if it stayed on your strong hand, it counted "
            "as dribbling rather than as crossovers."
        )
    elif rule == "same_hand" and (1 - share) < SAME_HAND_FLOOR:
        result.note(
            f"The ball changed hands on {share:.0%} of contacts. This one is "
            "meant to stay on the one hand, which is the whole reason it is "
            "worth doing."
        )


def summarise(reps: list[dict[str, Any]]) -> dict[str, Any]:
    """Per-session ball numbers worth keeping, for a coach and for scoring."""
    speeds = [float(r["speed"]) for r in reps if r.get("speed")]
    parts: dict[str, int] = {}
    for rep in reps:
        part = rep.get("part") or ""
        if part:
            parts[part] = parts.get(part, 0) + 1
    return {
        "contacts": len(reps),
        "median_speed": round(statistics.median(speeds), 3) if speeds else 0.0,
        "parts": parts,
    }


def _review_confirmation(
    spec: DrillSpec,
    reps: list[dict[str, Any]],
    track_quality: float | None,
    ball_contacts: int | None,
    ball_travel: float | None,
    result: BallReview,
) -> BallReview:
    """Confirm mode: the body counted the reps, the ball corroborates them.

    Deliberately asymmetric, and that asymmetry is the whole design. Not
    seeing a ball proves nothing -- a lacrosse ball is outside the detector's
    vocabulary, the light was bad, the phone was too far back -- so it never
    costs the athlete anything. Seeing the ball clearly and watching it never
    leave a hand while the arms threw forty times proves quite a lot.

    Penalising only on positive evidence is what stops this becoming a feature
    that quietly marks down every kid whose ball happens to be white.
    """
    if track_quality is None:
        # An older client, or one that never loaded the detector. Counts
        # exactly as it did before ball tracking existed.
        return result

    result.note(
        f"Ball visible in {result.quality:.0%} of frames."
    )
    if result.quality < spec.ball.min_track_quality:
        result.note(
            "Not enough to confirm a ball was used, which is common with a "
            "lacrosse ball, and it does not count against the session."
        )
        return result

    if len(reps) < CONFIRM_MIN_REPS or ball_contacts is None:
        return result

    share = ball_contacts / len(reps)
    if share < CONFIRM_MIN_SHARE:
        result.flag(
            f"The ball was tracked clearly for this session but was involved "
            f"in only {share:.0%} of the {len(reps)} throws counted."
        )
        return result

    # Contacts alone cannot tell a throw from a wave, because an arm whipping
    # forward with the ball still in it produces the same impulse beside the
    # same wrist. Whether the ball ever left the hand can.
    if ball_travel is not None and ball_travel < CONFIRM_MIN_TRAVEL:
        result.flag(
            "The ball never travelled away from your hands, so these look like "
            "throwing motions rather than throws."
        )
    return result
